authcrypt/main.cpp

```diff
@@ -53,132 +39,174 @@ static const unsigned char secret_key[16] = {
-    mbedtls_ctr_drbg_random(&drbg, ciphertext, nonce_len);
-
-    size_t ciphertext_len = 0;
-    /* Go for a conservative 16-byte (128-bit) tag
-     * and append it to the ciphertext */
-    const size_t tag_len = 16;
-    ret = mbedtls_cipher_auth_encrypt(&ctx, ciphertext, nonce_len,
-                              (const unsigned char *) metadata, sizeof metadata,
-                              (const unsigned char *) message, sizeof message,
-                              ciphertext + nonce_len, &ciphertext_len,
-                              ciphertext + nonce_len + sizeof message, tag_len );
-    if (ret != 0) {
-        mbedtls_printf("mbedtls_cipher_auth_encrypt() returned -0x%04X\r\n", -ret);
-        return 1;
-    }
-    ciphertext_len += nonce_len + tag_len;
-
-    /*
-     * The following information should now be transmitted:
-     * - first ciphertext_len bytes of ciphertext buffer
-     * - metadata if not already transmitted elsewhere
-     */
-    print_hex("ciphertext", ciphertext, ciphertext_len);
-
-    /*
-     * Decrypt-authenticate
-     */
-    unsigned char decrypted[128] = { 0 };
-    size_t decrypted_len = 0;
-
-    ret = mbedtls_cipher_setkey(&ctx, secret_key, 8 * sizeof secret_key, MBEDTLS_DECRYPT);
-    if (ret != 0) {
-        mbedtls_printf("mbedtls_cipher_setkey() returned -0x%04X\r\n", -ret);
-        return 1;
-    }
+private:
+    void print_hex(const char *title, const unsigned char buf[], size_t len)
+    {
+        mbedtls_printf("%s: ", title);
 
-    ret = mbedtls_cipher_auth_decrypt(&ctx,
-                              ciphertext, nonce_len,
-                              (const unsigned char *) metadata, sizeof metadata,
-                              ciphertext + nonce_len, ciphertext_len - nonce_len - tag_len,
-                              decrypted, &decrypted_len,
-                              ciphertext + ciphertext_len - tag_len, tag_len );
-    /* Checking the return code is CRITICAL for security here */
-    if (ret == MBEDTLS_ERR_CIPHER_AUTH_FAILED) {
-        mbedtls_printf("Something bad is happening! Data is not authentic!\r\n");
-        return 1;
-    }
-    if (ret != 0) {
-        mbedtls_printf("mbedtls_cipher_authdecrypt() returned -0x%04X\r\n", -ret);
-        return 1;
+        for (size_t i = 0; i < len; i++)
+            mbedtls_printf("%02x", buf[i]);
+
+        mbedtls_printf("\r\n");
     }
 
-    print_hex("decrypted", decrypted, decrypted_len);
+    /* Ciphertext buffer large enough to hold message + nonce + tag */
+    unsigned char ciphertext[128];
+    unsigned char decrypted[128];
 
-    mbedtls_printf("\r\nDONE\r\n");
+    /* Entropy pool for seeding PRNG */
+    mbedtls_entropy_context entropy;
 
-    return 0;
-}
+    /* Pseudo-random generator */
+    mbedtls_ctr_drbg_context drbg;
+    mbedtls_cipher_context_t cipher;
+};
 
 int main() {
-    int ret = example();
-    if (ret != 0) {
+    int ret;
+    Authcrypt *authcrypt = new Authcrypt();
```

Comment:
check if the heap is contigous. If not, the `ciphertext` and `decrypted` will not be properly read by HW acceleration engines